activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb

```diff
@@ -48,7 +48,13 @@ def execute(sql, name = nil, allow_retry: false)
 
           with_raw_connection(allow_retry: allow_retry) do |conn|
             log(sql, name) do
-              conn.async_exec(sql)
+              @warning_messages = []
+              conn.set_notice_processor { |message| @warning_messages << message } unless ActiveRecord.db_warnings_action == :ignore
```

Comment:
@paarthmadan and I are looking for feedback about setting the notice processor for PG, which is necessary to record warnings. Here, we're setting it on every query, which has potential performance implications. The alternative is to set the block for the notice processor _once_ (possibly when the PostgreSQL adapter is being initialized). Setting it during initialization minimizes overhead, but assumes that the `db_warnings_action` config isn't changing during runtime. 

We're opting _not_ to set the notice processor if `db_warnings_action` is `:ignore`, because the application might be relying on pre-existing behaviour from a notice processor block, and we don't want to clobber this if they're opting out of the new MySQL warnings feature. 

As long as we can assume that `db_warnings_action` isn't going to change during runtime, it should be fine to set it once, but if this config can change, then only setting the notice processor once will result in unintended behaviour (and we'll need to set it on every query).

Reply:
> As long as we can assume that `db_warnings_action` isn't going to change during runtime

I think we can assume that.